File: src/jray/model/MinimalTriangle.java
package jray.model;

import jray.common.Matrix4;
import jray.common.Object3D;
import jray.common.Ray;
import jray.common.Sphere;
import jray.common.Vect3;
import jray.math.Triangle;
import jray.math.Vect;
import jray.math.intersections.RayTriangle;

public class MinimalTriangle extends Object3D{

	protected Vect3 v2;
	protected Vect3 v3;
	protected Sphere bounds;
	
	public MinimalTriangle(Vect3 normal, Vect3 v1, Vect3 v2, Vect3 v3) {
		super(v1, normal);
		this.v2 = v2;
		this.v3 = v3;
		
		Vect3 avg = getBoundingSphereCenter();
		Vect3 tmp = new Vect3();
		Vect.subtract(avg, v3, tmp);
    	
		bounds = new Sphere(avg,tmp.length(),0xFF000000);
	}

	@Override
	public boolean contains(Vect3 hitPoint) {
		throw new RuntimeException("not implemented");
	}

	@Override
	public double getHitPointDistance(Ray r) {
		Vect3 v1v2 = new Vect3();
		Vect3 v1v3 = new Vect3();
		
		Vect.subtract(v2, position, v1v2);
		Vect.subtract(v3, position, v1v3);
		
		double ret = RayTriangle.getHitPointRayTriangleDistance(r.getOrigin(), r.getDirection(), position, v1v2, v1v3);
        if (ret <= 0) {
            return Double.POSITIVE_INFINITY;
        }
        return ret;
	}

	@Override
	public void getNormalAt(Vect3 hitPoint, Vect3 normal) {
		lookAt.copyDataTo(normal);
	}

	@Override
	public void rotate(Matrix4 rotationMatrix) {
		throw new RuntimeException("not implemented");
	}
	
    public Sphere getBoundingSphere(){
    	return bounds;
    }
	
	@Override
    public Vect3 getBoundingSphereCenter(){
		return Triangle.getCircumScribedCircleCenter(position, v2, v3);
    }
    
	@Override
    public double getBoundingSphereRadius(){
    	return bounds.getRadius();
    }
	
	public void purge(){
		bounds = null;
	}
}

File: JRayXLib/JRayXLib/Model/MinimalTriangle.cs
using System;
using JRayXLib.Colors;
using JRayXLib.Math;
using JRayXLib.Math.intersections;
using JRayXLib.Shapes;

namespace JRayXLib.Model
{
    public class MinimalTriangle : Basic3DObject
    {
        protected Sphere Bounds;
        protected Vect3 V2;
        protected Vect3 V3;

        public MinimalTriangle(Vect3 normal, Vect3 v1, Vect3 v2, Vect3 v3) : base(v1, normal)
        {
            V2 = v2;
            V3 = v3;

            Vect3 avg = GetBoundingSphereCenter();
            Vect3 tmp = avg - v3;

            Bounds = new Sphere(avg, tmp.Length(), Color.Black);
        }

        public override bool Contains(Vect3 hitPoint)
        {
            throw new Exception("not implemented");
        }

        public override double GetHitPointDistance(Shapes.Ray r)
        {
            Vect3 v1v2 = V2 - Position;
            Vect3 v1v3 = V3 - Position;

            double ret = RayTriangle.GetHitPointRayTriangleDistance(r.Origin, r.Direction, Position, v1v2, v1v3);
            if (ret <= 0)
            {
                return double.PositiveInfinity;
            }
            return ret;
        }

        public override Vect3 GetNormalAt(Vect3 hitPoint)
        {
            return LookAt;
        }

        public override void Rotate(Matrix4 rotationMatrix)
        {
            throw new Exception("not implemented");
        }

        public new Sphere GetBoundingSphere()
        {
            return Bounds;
        }

        public new Vect3 GetBoundingSphereCenter()
        {
            return Vect3Extensions.Avg(new[] {Position, V2, V3});
        }

        public override double GetBoundingSphereRadius()
        {
            return Bounds.GetRadius();
        }

        public void Purge()
        {
            Bounds = null;
        }
    }
}